File: go/rustcore_test.go
package canonicalize

import (
	"os"
	"runtime"
	"strings"
	"testing"
)

func TestNewRustCoreRequiresExplicitPath(t *testing.T) {
	if _, err := NewRustCore("libhtmltrust_canonicalization_ffi.so"); err == nil || !strings.Contains(err.Error(), "must be absolute") {
		t.Fatalf("relative path error: %v", err)
	}
	if _, err := NewRustCore("/definitely/missing/htmltrust-canonicalization-ffi.so"); err == nil {
		t.Fatal("NewRustCore unexpectedly succeeded")
	}
}

func TestNewRustCoreRejectsWrongABIFixture(t *testing.T) {
	if runtime.GOOS != "linux" {
		t.Skip("Linux shared-core fixture is not supported")
	}
	path := os.Getenv("HTMLTRUST_RUST_CORE_WRONG_ABI_LIB")
	if path == "" {
		t.Skip("HTMLTRUST_RUST_CORE_WRONG_ABI_LIB is not set")
	}
	if _, err := NewRustCore(path); err == nil || !strings.Contains(err.Error(), "unsupported htmltrust Rust core ABI version 999") {
		t.Fatalf("wrong ABI fixture error: %v", err)
	}
}

func TestNewRustCoreRejectsMissingOperationFixture(t *testing.T) {
	if runtime.GOOS != "linux" {
		t.Skip("Linux shared-core fixture is not supported")
	}
	path := os.Getenv("HTMLTRUST_RUST_CORE_MISSING_OPERATION_LIB")
	if path == "" {
		t.Skip("HTMLTRUST_RUST_CORE_MISSING_OPERATION_LIB is not set")
	}
	if _, err := NewRustCore(path); err == nil || !strings.Contains(err.Error(), "htmltrust Rust core symbol is missing: htmltrust_canonicalize_json_document_v1") {
		t.Fatalf("missing operation fixture error: %v", err)
	}
}

func TestRustCoreAllOperationsAndEdgeInputs(t *testing.T) {
	path := os.Getenv("HTMLTRUST_RUST_CORE_LIB")
	if path == "" {
		t.Skip("HTMLTRUST_RUST_CORE_LIB is not set")
	}
	core, err := NewRustCore(path)
	if err != nil {
		t.Fatal(err)
	}
	defer core.Close()
	if got, err := core.NormalizeText("A—B", false); err != nil || got != "A-B" {
		t.Fatalf("normalize: %q, %v", got, err)
	}
	if got, err := core.NormalizeText("a\x00b", false); err != nil || got != "a\x00b" {
		t.Fatalf("normalize embedded NUL: %q, %v", got, err)
	}
	if got, err := core.NormalizeText("", false); err != nil || got != "" {
		t.Fatalf("normalize empty: %q, %v", got, err)
	}
	base := "https://example.com/"
	if got, err := core.ExtractCanonicalText("<p>A</p>", false, &base); err != nil || got != "A" {
		t.Fatalf("extract: %q, %v", got, err)
	}
	emptyBase := ""
	if got, err := core.ExtractCanonicalText("<p>A</p>", false, &emptyBase); err != nil || got != "A" {
		t.Fatalf("extract with empty base: %q, %v", got, err)
	}
	if got, err := core.CanonicalizeClaims(map[string]string{"z": "2", "a": "1"}); err != nil || got != "a:1\nz:2\n" {
		t.Fatalf("claims: %q, %v", got, err)
	}
	if got, err := core.CanonicalizeClaims(nil); err != nil || got != "" {
		t.Fatalf("empty claims: %q, %v", got, err)
	}
	if _, err := core.CanonicalizeClaims(map[string]string{"bad": string([]byte{0xff})}); err == nil || !strings.Contains(err.Error(), "claim-malformed") {
		t.Fatalf("invalid UTF-8 claims error: %v", err)
	}
	if got, err := core.CanonicalizeJSONDocument([]byte(`{"z":0,"a":1}`)); err != nil || string(got) != `{"a":1,"z":0}` {
		t.Fatalf("jcs: %q, %v", got, err)
	}
	if _, err := core.CanonicalizeJSONDocument([]byte("{")); err == nil || err.Error() != "jcs-invalid-json" {
		t.Fatalf("invalid jcs error: %v", err)
	}
	if _, err := core.CanonicalizeJSONDocument([]byte(`"\uD800"`)); err == nil || !strings.Contains(err.Error(), "jcs-invalid-surrogate") {
		t.Fatalf("lone surrogate jcs error: %v", err)
	}
}

File: php/tests/RustCoreTest.php
<?php

declare(strict_types=1);

namespace HTMLTrust\Canonicalization\Tests;

use HTMLTrust\Canonicalization\RustCore;
use HTMLTrust\Canonicalization\RustCoreError;
use PHPUnit\Framework\TestCase;

final class RustCoreTest extends TestCase
{
    public function testPathMustBeExplicitAndNonEmpty(): void
    {
        $this->expectException(\InvalidArgumentException::class);
        new RustCore('');
    }

    public function testPathMustBeAbsolute(): void
    {
        $this->expectException(\InvalidArgumentException::class);
        $this->expectExceptionMessage('must be absolute');
        new RustCore('libhtmltrust_canonicalization_ffi.so');
    }

    public function testConstructorRejectsWrongAbiFixture(): void
    {
        if (PHP_OS_FAMILY !== 'Linux' || !class_exists('FFI')) {
            $this->markTestSkipped('Linux PHP FFI fixture is not supported');
        }
        $path = getenv('HTMLTRUST_RUST_CORE_WRONG_ABI_LIB');
        if (!is_string($path) || $path === '') {
            $this->markTestSkipped('HTMLTRUST_RUST_CORE_WRONG_ABI_LIB is not set');
        }
        $this->expectExceptionMessage('unsupported htmltrust Rust core ABI version 999');
        new RustCore($path);
    }

    public function testConstructorRejectsMissingOperationFixture(): void
    {
        if (PHP_OS_FAMILY !== 'Linux' || !class_exists('FFI')) {
            $this->markTestSkipped('Linux PHP FFI fixture is not supported');
        }
        $path = getenv('HTMLTRUST_RUST_CORE_MISSING_OPERATION_LIB');
        if (!is_string($path) || $path === '') {
            $this->markTestSkipped('HTMLTRUST_RUST_CORE_MISSING_OPERATION_LIB is not set');
        }
        $this->expectExceptionMessage('htmltrust_canonicalize_json_document_v1');
        new RustCore($path);
    }

    public function testAllOperationsAndEdgeInputs(): void
    {
        if (!class_exists('FFI')) {
            $this->markTestSkipped('ext-ffi is not installed');
        }
        $path = getenv('HTMLTRUST_RUST_CORE_LIB');
        if (!is_string($path) || $path === '') {
            $this->markTestSkipped('HTMLTRUST_RUST_CORE_LIB is not set');
        }
        $core = new RustCore($path);
        self::assertSame('A-B', $core->normalizeText('A—B'));
        self::assertSame("a\0b", $core->normalizeText("a\0b"));
        self::assertSame('', $core->normalizeText(''));
        self::assertSame('A', $core->extractCanonicalText('<p>A</p>', false, null));
        self::assertSame('A', $core->extractCanonicalText('<p>A</p>', false, ''));
        self::assertSame('A', $core->extractCanonicalText('<p>A</p>', false, 'https://example.com/'));
        self::assertSame("a:1\nz:2\n", $core->canonicalizeClaims(['z' => '2', 'a' => '1']));
        self::assertSame('', $core->canonicalizeClaims([]));
        self::assertSame('{"a":1,"z":0}', $core->canonicalizeJsonDocument('{"z":0,"a":1}'));
    }

    public function testClaimsNumericNameIsEncodedAsAnObject(): void
    {
        if (!class_exists('FFI')) {
            $this->markTestSkipped('ext-ffi is not installed');
        }
        $path = getenv('HTMLTRUST_RUST_CORE_LIB');
        if (!is_string($path) || $path === '') {
            $this->markTestSkipped('HTMLTRUST_RUST_CORE_LIB is not set');
        }
        $core = new RustCore($path);
        self::assertSame("0:value\n", $core->canonicalizeClaims(['0' => 'value']));
    }

    public function testInvalidJcsErrorIsMapped(): void
    {
        if (!class_exists('FFI')) {
            $this->markTestSkipped('ext-ffi is not installed');
        }
        $path = getenv('HTMLTRUST_RUST_CORE_LIB');
        if (!is_string($path) || $path === '') {
            $this->markTestSkipped('HTMLTRUST_RUST_CORE_LIB is not set');
        }
        $core = new RustCore($path);
        $this->expectException(RustCoreError::class);
        $this->expectExceptionMessage('jcs-invalid-json');
        $core->canonicalizeJsonDocument('{');
    }

    public function testLoneSurrogateJcsErrorIsMapped(): void
    {
        if (!class_exists('FFI')) {
            $this->markTestSkipped('ext-ffi is not installed');
        }
        $path = getenv('HTMLTRUST_RUST_CORE_LIB');
        if (!is_string($path) || $path === '') {
            $this->markTestSkipped('HTMLTRUST_RUST_CORE_LIB is not set');
        }
        $core = new RustCore($path);
        $this->expectException(RustCoreError::class);
        $this->expectExceptionMessage('jcs-invalid-surrogate');
        $core->canonicalizeJsonDocument('"\\uD800"');
    }
}
